File: baekjoon/dynamic/4883/main.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <vector>

using namespace std;
const int oo = 0x3f3f3f3f;

int cache[100001][3];
int N;
int map[100000][3];
int solve(int row, int col);
int main()
{
    int tc=0;
    while(1) {
        scanf("%d", &N);
        if(N == 0) break;
        memset(cache, -1, sizeof(cache));
        for(int i=0; i<N; i++) {
            for(int j=0; j<3; j++) {
                scanf("%d", &map[i][j]);
            }
        }
        printf("%d. %d\n", ++tc, solve(0,1));
    }
    return 0;
}

int solve(int row, int col)
{
    if(row == N-1) {
        if(col == 0) return map[row][col]+map[row][col+1];
        else if(col == 1) return map[row][col];
        else return oo;
    }
    
    int& ret = cache[row][col];
    if(ret != -1) return ret;
    ret = oo;
    
    if(row+1 < N) {
        ret = min(ret, solve(row+1, col)+map[row][col]);
    }
    if(row+1 < N && col+1 < 3) {
        ret = min(ret ,solve(row+1, col+1)+map[row][col]);
    }
    if(row+1 < N && col-1 > -1) {
        ret = min(ret ,solve(row+1, col-1)+map[row][col]);
    }
    
    if(col+1 < 3) {
        ret = min(ret, solve(row, col+1)+map[row][col]);
    } 
    
    return ret;
}

File: baekjoon/dynamic/4883/Main.java
import java.util.*;

class Main {
    static final int oo = 0x3f3f3f3f;
    static int N;
    static int dp[][];
    static int map[][];
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int tc=0;
        while(true) {
            N = sc.nextInt();
            if(N == 0) break;
            map = new int[100001][4];
            dp = new int[100001][4];
            for(int i=0; i<N; i++) {
                for(int j=0; j<3; j++) {
                    map[i][j] = sc.nextInt();
                }
            }
            
            dp[0][0] = oo;
            dp[0][1] = map[0][1];
            dp[0][2] = map[0][1]+map[0][2];
            
            for(int i=1; i<N; i++) {
                dp[i][0] = min(dp[i-1][0], dp[i-1][1])+map[i][0];
                dp[i][1] = min(dp[i-1][0], dp[i-1][1]);
                dp[i][1] = min(dp[i][1], dp[i-1][2]);
                dp[i][1] = min(dp[i][1], dp[i][0])+map[i][1];
                dp[i][2] = min(dp[i-1][1], dp[i-1][2]);
                dp[i][2] = min(dp[i][2], dp[i][1])+map[i][2];
            }
            
            System.out.printf("%d. %d\n", ++tc, dp[N-1][1]);
        }
    }
    
    public static int min(int a, int b) {
        return a>b?b:a;
    }
}